File: Training/1.1 Dabble/1.1.2-gift1.cpp
/*
ID:   andy1li
LANG: C++ 
TASK: gift1
*/

#include <bits/stdc++.h>
using namespace std;

#define range(stop) for (size_t i=0; i<stop; ++i)

/*----------------------------------------------------------------------------*/

void solve() {
    int n; cin >> n;

    vector<string> names(n);
    map<string, int> balance;
    range(n) {
        string name; cin >> name;
        names[i] = name;
        balance[name] = 0;
    }

    range(n) {
        string giver; cin >> giver;
        int amount, numReceivers; cin >> amount >> numReceivers;

        range(numReceivers) {
            string receiver; cin >> receiver;
            balance[giver]    -= amount / numReceivers;
            balance[receiver] += amount / numReceivers;
        }
    }
    
    for (string name: names) {
        cout << name << ' ' << balance[name] << endl;
    }
}

/*----------------------------------------------------------------------------*/

int main() {
    freopen("gift1.in",  "r", stdin);
    freopen("gift1.out", "w", stdout);
    solve();
}

File: Training/1.1 Dabble/1.1.2-gift1.py
'''
ID:   andy1li
LANG: PYTHON3
TASK: gift1
'''

def solve():
    n = int(input())
    balance = { input(): 0 for _ in range(n) }

    for _ in range(n):
        giver = input()
        amount, num_receivers = map(int, input().split())

        for r in range(num_receivers):
            balance[giver]   -= amount // num_receivers
            balance[input()] += amount // num_receivers

    for name, amount in balance.items():
        print_line(f'{name.strip()} {amount}')

#------------------------------------------------------------------------------#

def print_line(line): print(line); f_out.write( f'{line}\n' )

task = 'gift1'
with open(task+'.in') as f_in, open(task+'.out', 'w') as f_out:
    input = f_in.readline
    solve()